twenty-first/src/util_types/merkle_tree.rs

```diff
@@ -251,59 +276,75 @@ impl MerkleTree {
-        let auth_structure = indices.map(|idx| self.nodes[idx]).collect();
+        let auth_structure = indices.map(|idx| self.node(idx).unwrap()).collect();
         Ok(auth_structure)
     }
 
     pub fn root(&self) -> Digest {
         self.nodes[Self::ROOT_INDEX]
     }
 
-    pub fn num_leafs(&self) -> usize {
-        let node_count = self.nodes.len();
+    pub fn num_leafs(&self) -> MerkleTreeLeafIndex {
+        let node_count = MerkleTreeNodeIndex::try_from(self.nodes.len()).expect("usize to u64");
         debug_assert!(node_count.is_power_of_two());
         node_count / 2
     }
 
-    pub fn height(&self) -> usize {
+    pub fn height(&self) -> MerkleTreeHeight {
         let leaf_count = self.num_leafs();
         debug_assert!(leaf_count.is_power_of_two());
-        leaf_count.ilog2() as usize
+        MerkleTreeHeight::try_from(leaf_count.ilog2()).expect("log of num leafs should fit in u32")
     }
 
     /// All nodes of the Merkle tree.
-    pub fn nodes(&self) -> &[Digest] {
-        &self.nodes
+    pub fn nodes(&self) -> impl Iterator<Item = &Digest> {
+        self.nodes.iter()
     }
 
     /// The node at the given node index, if it exists.
-    pub fn node(&self, index: usize) -> Option<Digest> {
-        self.nodes.get(index).copied()
+    pub fn node(&self, index: MerkleTreeNodeIndex) -> Option<Digest> {
+        usize::try_from(index)
+            .ok()
+            .and_then(|idx| self.nodes.get(idx).copied())
     }
 
     /// All leafs of the Merkle tree.
-    pub fn leafs(&self) -> &[Digest] {
-        let first_leaf = self.nodes.len() / 2;
-        &self.nodes[first_leaf..]
+    pub fn leafs(&self) -> impl Iterator<Item = &Digest> {
+        self.nodes.iter().skip(
+            (self.num_nodes() / 2)
+                .try_into()
+                .expect("MerkleTreeNodeIndex to usize conversion error"),
+        )
     }
 
     /// The leaf at the given index, if it exists.
-    pub fn leaf(&self, index: usize) -> Option<Digest> {
-        let first_leaf_index = self.nodes.len() / 2;
-        self.nodes.get(first_leaf_index + index).copied()
+    pub fn leaf(&self, index: MerkleTreeLeafIndex) -> Option<Digest> {
+        let first_leaf_index = self.num_nodes() / 2;
+        self.node(first_leaf_index + index)
     }
 
-    pub fn indexed_leafs(&self, indices: &[usize]) -> Result<Vec<(usize, Digest)>> {
+    pub fn indexed_leafs(
+        &self,
+        indices: &[MerkleTreeLeafIndex],
+    ) -> Result<Vec<(MerkleTreeLeafIndex, Digest)>> {
         let num_leafs = self.num_leafs();
         let invalid_index = MerkleTreeError::LeafIndexInvalid { num_leafs };
         let maybe_indexed_leaf = |&i| self.leaf(i).ok_or(invalid_index).map(|leaf| (i, leaf));
 
         indices.iter().map(maybe_indexed_leaf).collect()
     }
 
+    fn num_nodes(&self) -> MerkleTreeNodeIndex {
+        MerkleTreeNodeIndex::try_from(self.nodes.len())
+            .expect("`MerkleTreeNodeIndex` should be large enough to hold any usize")
```

Comment:
Maybe use the existing `err::USIZE_TO_U64_ERR` string? Here, I'm less sure.

Reply:
I don't see why not. More specifically, I don't see why we should treat the two cases where this conversion happens differently.